modules/ensemble/lib/action/moengage_actions.dart

```diff
@@ -0,0 +1,382 @@
+import 'package:ensemble/framework/action.dart';
+import 'package:ensemble/framework/error_handling.dart';
+import 'package:ensemble/framework/event.dart';
+import 'package:ensemble/framework/extensions.dart';
+import 'package:ensemble/framework/scope.dart';
+import 'package:ensemble/framework/stub/moengage_manager.dart';
+import 'package:ensemble/screen_controller.dart';
+import 'package:ensemble/util/utils.dart';
+import 'package:ensemble_ts_interpreter/invokables/invokable.dart';
+import 'package:flutter/material.dart';
+import 'package:get_it/get_it.dart';
+import 'package:moengage_flutter/moengage_flutter.dart';
+
+// Main action types available for MoEngage
+enum MoEngageActionType {
+  // User attributes
+  setUser, // Generic action to set any user attribute
+
+  // Events
+  trackEvent,
+
+  // App State
+  appConfig,
+
+  // Push/Notification
+  pushConfig,
+
+  // InApp & Cards
+  displayConfig
+}
+
+// Sub-types for setUser action
+enum MoEngageUserAttributeType {
+  uniqueId,
+  userName,
+  firstName,
+  lastName,
+  email,
+  mobile,
+  gender,
+  birthday,
+  location,
+  alias,
+  custom, // For custom attribute with key-value
+  timestamp,
+  locationAttribute
+}
+
+// Sub-types for app configuration
+enum MoEngageAppConfigType {
+  enableSdk,
+  disableSdk,
+  enableDataTracking,
+  disableDataTracking,
+  enableDeviceIdTracking,
+  disableDeviceIdTracking,
+  enableAndroidIdTracking,
+  disableAndroidIdTracking,
+  enableAdIdTracking,
+  disableAdIdTracking,
+  setAppStatus,
+  logout,
+  deleteUser
+}
+
+// Sub-types for push notification configuration
+enum MoEngagePushConfigType {
+  registerForPush,
+  registerForProvisionalPush,
+  passFCMToken,
+  passPushKitToken,
+  passFCMPushPayload,
+  requestPushPermission,
+  updatePermissionCount,
+  pushPermissionResponse
+}
+
+// Sub-types for display configuration
+enum MoEngageDisplayConfigType {
+  showInApp,
+  showNudge,
+  setContext,
+  resetContext
+}
+
+class MoEngageAction extends EnsembleAction {
+  final MoEngageActionType? type;
+  final dynamic actionType; // Sub type enum based on main type
+  final dynamic value;
+  final Map<String, dynamic>? properties; // For events and custom attributes
+  final String? attributeKey; // For custom user attributes
+  final EnsembleAction? onSuccess;
+  final EnsembleAction? onError;
+
+  MoEngageAction(
+      {super.initiator,
+      required this.type,
+      this.actionType,
+      this.value,
+      this.properties,
+      this.attributeKey,
+      this.onSuccess,
+      this.onError});
+
+  factory MoEngageAction.fromYaml({Invokable? initiator, Map? payload}) {
+    if (payload == null || payload['type'] == null) {
+      throw ConfigError('MoEngage action requires type');
+    }
+
+    return MoEngageAction(
+        initiator: initiator,
+        type: MoEngageActionType.values.from(payload['type']) ?? null,
```

Comment:
Since type is mandatory make it non-null by moving `MoEngageActionType.values.from(payload['type'])` above in payload condition check, so you can throw error in `MoEngageAction.fromYaml` rather than in `execute`